ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos/ChainState.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE StrictData #-}
+
+-- | Chain state for Transitional Praos
+module Ouroboros.Consensus.Protocol.TPraos.ChainState where
+
+import Cardano.Prelude (NoUnexpectedThunks(..))
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+import Ouroboros.Network.Block (SlotNo (..))
+import Ouroboros.Network.Point (WithOrigin (..))
+import Ouroboros.Consensus.Protocol.TPraos.Util
+
+import qualified STS.Prtcl as STS
+
+-- | Praos chain state.
+--
+-- The transitional praos implementation itself has a concept of state, given by
+-- the STS.State (PRTCL c) type. This chain state, however, doesn't support
+-- rewinding.
+--
+--
+data TPraosChainState c = TPraosChainState
+  { -- | Anchor
+    --
+    -- The anchor is the earliest slot to which we can roll back. It should
+    -- correspond to the first slot with an entry in the historical state.
+    --
+    -- We store this for easy computation of whether a rollback is allowed, and
+    -- for sanity checking of the state.
+    anchor :: WithOrigin SlotNo
+
+    -- | Historical state snapshots.
+  , historicalStates :: Map (WithOrigin SlotNo) (STS.State (PRTCL c))
```

Comment:
Add a comment considering whether it's okay to keep `k` states, particularly wrt the map of KES key issue numbers

Reply:
Add a comment to say we could consider making this a finger tree if `rewind` seems to be a hotspot